physical/dynamodb.go

```diff
@@ -481,57 +510,86 @@ func (l *DynamoDBLock) Value() (bool, string, error) {
 func (l *DynamoDBLock) tryToLock(stop, success chan struct{}, errors chan error) {
 	ticker := time.NewTicker(DynamoDBLockRetryInterval)
 
-	record := DynamoDBRecord{
-		Path:  recordPathForVaultKey(l.key),
-		Key:   recordKeyForVaultKey(l.key),
-		Value: []byte(l.value),
-	}
-	item, err := dynamodbattribute.ConvertToMap(record)
-	if err != nil {
-		errors <- err
-		return
-	}
-
 	for {
 		select {
 		case <-stop:
 			ticker.Stop()
 		case <-ticker.C:
-			_, err := l.backend.client.PutItem(&dynamodb.PutItemInput{
-				TableName:           aws.String(l.backend.table),
-				Item:                item,
-				ConditionExpression: aws.String("attribute_not_exists(#p) or attribute_not_exists(#k)"),
-				ExpressionAttributeNames: map[string]*string{
-					"#p": aws.String("Path"),
-					"#k": aws.String("Key"),
-				},
-			})
+			err := l.writeItem()
 			if err != nil {
-				if err, ok := err.(awserr.Error); ok && err.Code() != "ConditionalCheckFailedException" {
-					errors <- err
-				}
-				if l.recovery {
-					_, err := l.backend.client.DeleteItem(&dynamodb.DeleteItemInput{
-						TableName: aws.String(l.backend.table),
-						Key: map[string]*dynamodb.AttributeValue{
-							"Path": {S: aws.String(record.Path)},
-							"Key":  {S: aws.String(record.Key)},
-						},
-					})
-					if err != nil {
-						errors <- fmt.Errorf("could not delete lock record: %s", err)
-					} else {
-						l.recovery = false
+				if err, ok := err.(awserr.Error); ok {
+					// Don't report a condition check failure, this means that the lock
+					// is already being held.
+					if err.Code() != "ConditionalCheckFailedException" {
+						errors <- err
 					}
+				} else {
+					// Its not an AWS error, and is probably not transient, bail out.
+					errors <- err
+					return
 				}
 			} else {
 				ticker.Stop()
 				close(success)
+				return
 			}
 		}
 	}
 }
 
+func (l *DynamoDBLock) periodicallyRenewLock(done chan struct{}) {
+	for {
+		select {
+		case <-time.After(l.renewInterval):
```

Comment:
Nitpicking: You could use a single `time.Ticker` instead of calling `time.After` multiple times. `time.After` instantiates a `time.Timer`, so a `time.Ticker` *should* be a bit more efficient here. (Please correct me if I'm wrong)